[WebGPU] Add the ability to fuzz wgslc

litherum opened the pull request:
#### 751a8c03a46a2b18f88e4385e7c15a2dd49cbddf
<pre>
[WebGPU] Add the ability to fuzz wgslc
<a href="https://bugs.webkit.org/show_bug.cgi?id=259355">https://bugs.webkit.org/show_bug.cgi?id=259355</a>
rdar://112576959

Reviewed by David Kilzer.

The way this works is:
% set-webkit-configuration --debug --asan --libFuzzer
% cd Source/WebGPU
% make SCHEME=wgslc
% ASAN_OPTIONS=whatever DYLD_FRAMEWORK_PATH=/path/to/Products/Debug DYLD_LIBRARY_PATH=/path/to/Products/Debug /path/to/Products/Debug/wgslc

This patch adds a new configuration option, named &quot;libFuzzer&quot; to WebKit. It sets
the ENABLE_LIBFUZZER Xcode variable, which automatically adds -fsanitize=fuzzer to
compilations. It also sets the ENABLE_LIBFUZZER preprocessor macro, which we can
use to conditionally use LLVMFuzzerTestOneInput() instead of main() if fuzzing is
enabled. Enabling fuzzing also enables ASAN (because of course it does).

* Configurations/Sanitizers.xcconfig:
* Source/WebGPU/WGSL/wgslc.cpp:
(runWGSL):
(LLVMFuzzerTestOneInput):
* Tools/Scripts/set-webkit-configuration:
(printCurrentSettings):
* Tools/Scripts/webkitdirs.pm:
(determineLibFuzzerIsEnabled):
(libFuzzerIsEnabled):
(XcodeOptions):
(generateBuildSystemFromCMakeProject):

Canonical link: <a href="https://commits.webkit.org/266202@main">https://commits.webkit.org/266202@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/202831dc183675426d25951f7f9517c6045591b9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13196 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13509 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13843 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14934 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12564 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13264 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16017 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13536 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15259 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13363 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14035 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11155 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15492 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11323 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11909 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18963 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/11226 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12398 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12076 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15285 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/12487 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12564 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/10432 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/13244 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11837 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3482 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3231 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16159 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/13625 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12407 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3277 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

litherum commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -1219,9 +1235,10 @@ sub XcodeOptions
     }
     push @options, ("-configuration", $configuration);
-    push @options, ("ENABLE_ADDRESS_SANITIZER=YES") if $asanIsEnabled;
+    push @options, ("ENABLE_ADDRESS_SANITIZER=YES") if $asanIsEnabled || $libfuzzerIsEnabled;
```
Feedback: Let's make the 2 settings independent!

litherum commented on Configurations/Sanitizers.xcconfig:
```diff
@@ -82,3 +82,7 @@ WK_XCODE_VERSION_AFTER_13_3_1700 = YES;
+
+// Libfuzzer
+
+// Nothing extra is necessary for libfuzzer!
```
This isn't right. We need:

```
WK_LIBFUZZER_OTHER_CFLAGS_YES = -fsanitize-coverage=inline-8bit-counters,trace-cmp -fno-sanitize-coverage=pc-table -Wno-error;
```

(and maybe something similar for ldflags too?)

ddkilzer commented on Configurations/Sanitizers.xcconfig:
```diff
@@ -82,3 +82,9 @@ WK_XCODE_VERSION_AFTER_13_3_1700 = YES;
+
+WK_LIBFUZZER_OTHER_CFLAGS_YES = -fsanitize-coverage=inline-8bit-counters,trace-cmp -fno-sanitize-coverage=pc-table -Wno-error;
+
+WK_LIBFUZZER_OTHER_CPLUSPLUSFLAGS_YES = ;
```
You can just omit this unless you want to leave it as a placeholder.  (Not required to change this to land the patch.)

litherum replied:
I'd like to leave it as a placeholder, if that's okay.

ddkilzer commented on Source/WebGPU/WGSL/wgslc.cpp:
```diff
@@ -143,3 +144,23 @@ int main(int argc, char** argv)
+    }
+
+    return 0;
+}
```
I expected this code to appear in the `#else ... #endif // ENABLE(LIBFUZZER)` section of the macro.

Technically, it will be compiled and then removed via dead code stripping, but I think it would be best if it appeared in an  `#else ... #endif // ENABLE(LIBFUZZER)` section.

An exception would be if there is a case where we want to enable this code without compiling with `ENABLE(LIBFUZZER)` being true, but I don't know of a use case there.

ddkilzer commented on Tools/Scripts/set-webkit-configuration:
```diff
@@ -44,6 +44,7 @@ Usage: $programName [options]
   --[no-]asan                        Enable or disable clang address sanitizer
   --[no-]tsan                        Enable or disable clang thread sanitizer
   --[no-]ubsan                       Enable or disable clang undefined behavior sanitizer
+  --[no-]libfuzzer                   Enable or disable clang libfuzzer
```
The name of the library is `libFuzzer`, so my OCD really wants just the description fixed (the `--libfuzzer` switch can remain all lowercase).

However, I don't think this needs to be fixed before landing the PR.

litherum replied:
Oh really? I thought it was “libfuzzer”

litherum replied:
You're right! https://www.llvm.org/docs/LibFuzzer.html

ddkilzer reviewed (approved):
r=me

We can land as-is, although I'd prefer the `LLVMFuzzerTestOneInput` declaration and implementation inside the `#else ... #endif // ENABLE(LIBFUZZER)` macros.

litherum committed:
[WebGPU] Add the ability to fuzz wgslc
https://bugs.webkit.org/show_bug.cgi?id=259355
rdar://112576959

Reviewed by David Kilzer.

The way this works is:
% set-webkit-configuration --debug --asan --libFuzzer
% cd Source/WebGPU
% make SCHEME=wgslc
% ASAN_OPTIONS=whatever DYLD_FRAMEWORK_PATH=/path/to/Products/Debug DYLD_LIBRARY_PATH=/path/to/Products/Debug /path/to/Products/Debug/wgslc

This patch adds a new configuration option, named "libFuzzer" to WebKit. It sets
the ENABLE_LIBFUZZER Xcode variable, which automatically adds -fsanitize=fuzzer to
compilations. It also sets the ENABLE_LIBFUZZER preprocessor macro, which we can
use to conditionally use LLVMFuzzerTestOneInput() instead of main() if fuzzing is
enabled. Enabling fuzzing also enables ASAN (because of course it does).

* Configurations/Sanitizers.xcconfig:
* Source/WebGPU/WGSL/wgslc.cpp:
(runWGSL):
(LLVMFuzzerTestOneInput):
* Tools/Scripts/set-webkit-configuration:
(printCurrentSettings):
* Tools/Scripts/webkitdirs.pm:
(determineLibFuzzerIsEnabled):
(libFuzzerIsEnabled):
(XcodeOptions):
(generateBuildSystemFromCMakeProject):

Canonical link: https://commits.webkit.org/266202@main
